### PR TITLE
MST: use response from transport

### PR DESCRIPTION
**Pull request by lebdron**
### Description of the Change
- Add response callback parameter to AsyncGrpcClient. Now it is possible to process responses for asynchronous network calls.
- Add response status handling to MST transport. Allows not to assume that the diff was received as in https://github.com/hyperledger/iroha/pull/362 , but to make a decision based on the status instead.
It is also possible to implement an alternative approach to https://github.com/hyperledger/iroha/pull/364 based on an alternative response structure.

### Benefits
Ability to process responses for asynchronous network calls

### Possible Drawbacks 
None

### CODE REVIEW

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 1]
```diff
@@ -140,30 +144,43 @@ void MstTransportGrpc::subscribe(
...
+    log_->info("Propagate MstState to peer {}", to.address());
+    sendStateAsyncImpl(to,
+                       providing_state,
+                       my_key_,
+                       *async_call_,
+                       [s](auto &status, auto &) {
+                         s.on_next(status.ok());
+                         s.on_completed();
+                       },
+                       sender_factory_.value_or(default_sender_factory));
```
Comment (MBoldyrev) [inline irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp:159]:
```suggestion
                       },
                       sender_factory_.value_or(default_sender_factory));
    s.on_completed();
```

File: test/module/irohad/multi_sig_transactions/mst_processor_test.cpp [hunk 2]
```diff
@@ -294,6 +295,60 @@ TEST_F(MstProcessorTest, onNewPropagationUsecase) {
...
+ *
+ * @then diff is not applied to storage
+ */
+TEST_F(MstProcessorTest, SendStateFailure) {
+  // ---------------------------------| given |---------------------------------
+  auto quorum = 2u;
+  mst_processor->propagateBatch(addSignaturesFromKeyPairs(
+      makeTestBatch(txBuilder(1, time_after, quorum)), 0, makeKey()));
+  EXPECT_CALL(*transport, sendState(_, _))
+      .WillOnce(Return(rxcpp::observable<>::just(false)));
```
Comment (MBoldyrev) [inline test/module/irohad/multi_sig_transactions/mst_processor_test.cpp:340]: should there be also be a test with empty observable? in my understanding of the modular test approach it should. I'd rewrite these tests with a fixture then, to have only the `sendState` expectation and state diff check in each `TEST_F`

File: test/module/shared_model/interface_mocks.hpp [hunk 3]
```diff
@@ -190,22 +190,22 @@ struct MockPeer : public shared_model::interface::Peer {
 };
 
 inline auto makePeer(
-    const std::string &address,
-    const std::string &pub_key,
+    std::string_view address,
+    std::string_view pub_key,
```
Comment (MBoldyrev) [inline test/module/shared_model/interface_mocks.hpp:194]:
```suggestion
    PublicKeyHexStringView pub_key,
```

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp (hunk 1 above)
Comment (lebdron) [reply irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp:159]: Discussed offline, `on_completed` should not be called right after making an asynchronous call.

File: test/module/irohad/multi_sig_transactions/mst_processor_test.cpp (hunk 2 above)
Comment (lebdron) [reply test/module/irohad/multi_sig_transactions/mst_processor_test.cpp:340]: Discussed offline, transport interface is restricted to return exactly one value at the moment, and if the interface changes, then the test will be refactored.
